lib/keyverificationsession.cpp

```diff
@@ -0,0 +1,522 @@
+            return;
+        } else {
+            startSentByUs = false;
+        }
+    }
+    QByteArray publicKey(olm_sas_pubkey_length(m_sas), '\0');
+    olm_sas_get_pubkey(m_sas, publicKey.data(), publicKey.size());
+    const auto canonicalEvent = QString(QJsonDocument(event.contentJson()).toJson(QJsonDocument::Compact));
+    auto commitment = QString(QCryptographicHash::hash((QString(publicKey) % canonicalEvent).toLatin1(), QCryptographicHash::Sha256).toBase64());
+    commitment = commitment.left(commitment.indexOf('='));
+
+    auto acceptEvent = makeEvent<KeyVerificationAcceptEvent>(QJsonObject {
+        {"type", "m.key.verification.accept"},
+        {"content", QJsonObject {
+            {"commitment", commitment},
+            {"hash", "sha256"},
+            {"key_agreement_protocol", "curve25519-hkdf-sha256"},
+            {"message_authentication_code", "hkdf-hmac-sha256"},
+            {"method", "m.sas.v1"},
+            {"short_authentication_string", QJsonArray {
+                "decimal",
+                "emoji",
+            }},
+            {"transaction_id", m_transactionId},
+        }}
+    });
+    m_connection->sendToDevice(m_remoteUserId, m_remoteDeviceId, std::move(acceptEvent), m_encrypted);
+    setState(ACCEPTED);
+}
+
+void KeyVerificationSession::handleAccept(const KeyVerificationAcceptEvent& event)
+{
+    if(state() != WAITINGFORACCEPT) {
+        cancelVerification(UNEXPECTED_MESSAGE);
+        return;
+    }
+    m_commitment = event.commitment();
+    sendKey();
+    setState(WAITINGFORKEY);
+}
+
+void KeyVerificationSession::handleMac(const KeyVerificationMacEvent& event)
+{
+    QStringList keys = event.mac().keys();
+    keys.sort();
+    const auto& key = keys.join(",");
+    const QString edKeyId = "ed25519:"_ls % m_remoteDeviceId;
+
+    if (calculateMac(m_connection->edKeyForUserDevice(m_remoteUserId, m_remoteDeviceId), true, edKeyId) != event.mac()[edKeyId]) {
+        cancelVerification(KEY_MISMATCH);
+        return;
+    }
+
+    if (calculateMac(key, true) != event.keys()) {
+        cancelVerification(KEY_MISMATCH);
+        return;
+    }
+
+    m_connection->database()->setSessionVerified(edKeyId);
+    emit m_connection->sessionVerified(m_remoteUserId, m_remoteDeviceId);
+    macReceived = true;
+
+    if (state() == WAITINGFORMAC) {
+        setState(DONE);
+        sendDone();
+        emit finished();
+        deleteLater();
+    }
+}
+
+void KeyVerificationSession::handleDone(const KeyVerificationDoneEvent& event)
+{
+    if (state() != DONE) {
+        cancelVerification(UNEXPECTED_MESSAGE);
+    }
+}
+
+void KeyVerificationSession::handleCancel(const KeyVerificationCancelEvent& event)
+{
+    setError(stringToError(event.code()));
+    setState(CANCELED);
+}
+
+std::pair<QString, QString> KeyVerificationSession::emojiForCode(int code)
+{
+    static QJsonArray data;
+    if (data.isEmpty()) {
+        QFile dataFile(":/sas-emoji.json");
```

Comment:
I think it's better to keep this file out on the file system rather than swallow it in a Qt resource, so that it could be updated independently from the client (and possibly even automatically downloaded by the client whenever a new version is found).